### PR TITLE
Introduce PackageRef

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
this advance the state of #17995 

We propagate in ApiCommand and fix preprocessing

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala [hunk 1]
```diff
@@ -179,6 +179,9 @@ final case class ResultNeedUpgradeVerification[A](
 ) extends Result[A]
 
 object Result {
+
+  def done[A](a: A): Result[A] = ResultDone(a)
```
Comment (akrmn) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala:183]: is this ever used?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 2]
```diff
@@ -249,24 +250,49 @@ private[lf] final class CommandPreprocessor(
     speedy.Command.LookupByKey(templateId, key)
   }
 
+  private[preprocessing] def unsafeResolveTyConName(
+      pkgPriority: Map[Ref.PackageName, Ref.PackageId],
```
Comment (akrmn) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:254]: why is this argument called`pkgPriority` here and `pkgMap` elsewhere?

**Review by akrmn — APPROVED** [review]
Looks good, thank you @remyhaemmerle-da!

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:254]: good catch.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala:183]: I split the PR, it is a left over. 
I will remove it.

**Pull request merged**